### PR TITLE
Packet shenanigans

### PR DESCRIPTION
**Pull request by wingo**
This patch improves "snabb lwaftr bench" by 19%, is rock solid under a bare-metal lwaftr test with a big binding table, and lets us reach line rate with a virtualized lwaftr, though it is currently less solid than it could be.  PTAL, and fingers crossed for the CI!

### CODE REVIEW

**Comment by wingo** [comment]
Cc @lukego because I know you are always curious about these things :)

**Comment by dpino** [comment]
LGTM.

File: src/core/packet.lua [hunk 1]
```diff
@@ -59,23 +60,28 @@ end
...
    ffi.copy(p.data, ptr, len)                -- Fill the gap
-   p.length = p.length + len
    return p
 end
 
 -- Move packet data to the left. This shortens the packet by dropping
 -- the header bytes at the front.
 function shiftleft (p, bytes)
-   C.memmove(p.data, p.data+bytes, p.length-bytes)
+   -- C.memmove(p.data, p.data+bytes, p.length-bytes)
```
Comment (kbara) [inline src/core/packet.lua]: There should perhaps be a range check here (p.length >= bytes?)

File: src/core/packet.lua [hunk 2]
```diff
@@ -59,23 +60,28 @@ end
...
+   p.data = p.data + bytes
    p.length = p.length - bytes
 end
 
 -- Move packet data to the right. This leaves length bytes of data
 -- at the beginning of the packet.
 function shiftright (p, bytes)
-   C.memmove(p.data + bytes, p.data, p.length)
+   if p.data - bytes >= p.data_ - C.PACKET_HEADROOM_SIZE then
```
Comment (kbara) [inline src/core/packet.lua]: Perhaps there should be a range check here (p.length + bytes <= PACKET_PAYLOAD_SIZE ?)

**Comment by wingo** [comment]
Good comments @kbara, I added some assertions.

File: src/core/packet.lua [hunk 3]
```diff
@@ -58,24 +60,36 @@ end
...
    return p
 end
 
 -- Move packet data to the left. This shortens the packet by dropping
 -- the header bytes at the front.
 function shiftleft (p, bytes)
-   C.memmove(p.data, p.data+bytes, p.length-bytes)
+   assert(bytes >= 0 and bytes <= p.length)
+   p.data = p.data + bytes
+   p.headroom = p.headroom + bytes
```
Comment (kbara) [inline src/core/packet.lua:76]: Can't p.headroom become problematic (ie, if bytes > PACKET_HEADROOM_SIZE, either in one call or across several calls in total?)

Comment (kbara) [inline src/core/packet.lua:76]: (actually, this should be in terms of headroom)

**Comment by wingo** [comment]
Ah, one more thing then.  The headroom is 48 right now which is enough for a 40-byte IPv6 header.  However it's not enough for that _and_ a virtio header, which is 12 or 14 bytes.  But, in the lwaftr most packets come from virtio, so there will be enough -- but, some packets don't, like NDP packets or something.  Better to just reserve a few more bytes so that we know that the virtio interfaces won't have to memmove.

File: src/core/packet.lua (hunk 3 above)
Comment (wingo) [inline src/core/packet.lua:76]: The invariant is that p.headroom + p.length <= max_payload.  This routine takes bytes away from length and gives them to headroom.

Comment (kbara) [inline src/core/packet.lua:76]: So it is; nevermind. :-)

File: src/core/packet.lua [hunk 4]
```diff
@@ -58,24 +60,36 @@ end
...
 end
 
 -- Move packet data to the right. This leaves length bytes of data
 -- at the beginning of the packet.
 function shiftright (p, bytes)
-   C.memmove(p.data + bytes, p.data, p.length)
+   if bytes <= p.headroom then
+      -- Take from the headroom.
+      assert(bytes >= 0)
+      p.headroom = p.headroom - bytes
```
Comment (kbara) [inline src/core/packet.lua:86]: Maybe I'm being slow, but isn't this missing something like:
`p.data = p.data_ + p.headroom` ? That doesn't auto-update when headroom is changed.

Comment (kbara) [inline src/core/packet.lua:86]: Nevermind, that's outside of the if. :-)

**Comment by kbara** [comment]
LGTM

**Comment by wingo** [comment]
thank you!

**Comment by lukego** [comment]
@wingo This is exciting work :+1:. Getting a ~20% application-level speedup with such a simple change is very nice. This also seems perfectly timed for using the "matrix tests" from @domenkozar to look for benefits in other applications too.

Idea: How about if we could achieve the same effect while preserving the original simple definition of `struct packet`?

Specifically, I am imagining if we could keep the original definition but allocate packet memory in such a way that we can do the cheap/non-copying shift operation (with a slightly different API). If we would always allocate an extra (say) 256 bytes for each packet, and we would align the allocation such that we can infer the amount of headroom/tailroom available from the low bits of the address, then we could perhaps implement a fast `relocate()` operation to replace `shiftleft` and `shiftright`:

``` lua
-- packet.relocate(packet, offset) => packet'
function packet.relocate (p1, offset)
   -- Get the address of the original packet
   local addr1 = ffi.cast("uint8_t*", p1) -- pointer to original packet struct
   -- Calculate the address of the new packet struct.
   -- The address points to the same memory but at a different offset.
   local addr2 = addr1 + offset
   -- Check for headroom/tailroom overflow.
   -- (Packets need to be allocated with an alignment such that this test makes sense.)
   assert(band(addr1, 255) == band(addr1, 255), "packet displaced beyond reserved memory")
   -- Create the new packet struct, destructively reusing memory from the original.
   local p2 = ffi.cast("struct packet *", addr2)
   -- Copy the header fields (well, just length) to their new locations.
   p2.length = p1.length -- XXX barrier needed? aliasing risks? (These fields may overlap in memory)
   return p2
end
```

Generally I am excited about the idea of having enough performance test coverage in the upstream CI that we can make these kinds of radical changes with confidence & updating all relevant applications at the same time if needed.

Looking forward to kicking the tires on this when a version comes upstream :)

**Comment by wingo** [comment]
@lukego interesting idea!  I wonder though, once you start to consider relocating packets, what is the advantage of:

```
struct packet { uint8_t data[10*1024]; uint16_t length; }
```

over

```
struct packet { uint8_t *data; uint16_t length; }
```

I think that given that the packet length is on another cache line from the packet data in the current configuration, that there would be the same number of cache accesses to get the indirected data from the pointer, compared with inline allocation.  But, definitely a question for the test matrix :)

**Comment by lukego** [comment]
@wingo Yes! On that second formulation I am also keen to compare passing `struct packet` by value rather than by reference. The struct is quite tiny - I suppose 16 bytes with default field alignment - so we could store it inline in `struct link` to avoid a pointer indirection.

**Pull request merged**